### PR TITLE
fix ctests on intel compiler 

### LINKED ISSUE #773: Some tests failing on orion with intel spack-stack
## Description
The following tests fail on orion with intel spack-stack:

- `test_soca_convertstate`:
> 1266: Test Val : 5.9782488939415996e+00
1266: Ref  Val : 5.9783431324160823e+00
1266: Delta    : 9.4238474482644108e-05
1266: Relative tolerance: 5.9782960131788415e-06
1266: Absolute tolerance: 0.0000000000000000e+00
1266: Test Line: '   tocn   min=-1.8880081892310945   max=31.7033530247322126   mean=5.9782488939415996'
1266: Ref Line : '   tocn   min=-1.8880081892310945   max=31.7033530247322091   mean=5.9783431324160823'

Should be fixable by small adjustment of tolerance.

- `test_soca_dirac_soca_mask` and `test_soca_dirac_soca_nomask`:
> 1312: Caught SIGFPE: floating-point divide by zero
1312: Caught SIGFPE: floating-point divide by zero
1312:  0# sigfpe_handler(int, siginfo_t*, void*) in /work/noaa/da/ashlyaev/jedi/build/bin/../lib64/liboops.so
1312:  1# 0x00007FFA747E15D0 in /lib64/libpthread.so.0
1312:  2# atlas::grid::spacing::gaussian::compute_gaussian_quadrature_npole_equator(unsigned long, double*, double*) in /work/noaa/da/role-da/spack-stack/spack-stack-v1/envs/skylab-1.0.0-intel-2022.0.2/install/intel/2022.0.2/ecmwf-atlas-0.29.0-k4yxszy/lib64/libatlas.so.0.29
1312:  3# atlas::grid::spacing::gaussian::gaussian_latitudes_npole_equator(unsigned long, double*) in /work/noaa/da/role-da/spack-stack/spack-stack-v1/envs/skylab-1.0.0-intel-2022.0.2/install/intel/2022.0.2/ecmwf-atlas-0.29.0-k4yxszy/lib64/libatlas.so.0.29

I remember having (I think) the same issue on my mac at some point. I debugged then a bit and attributed it to a compiler optimization gone wrong. I think we had a similar issue at some point with some of the ufo code (may have to do with vectorization and dividing by zero past the end of array or something like this). I'll try to find how we fixed that.

- `test_soca_3dvar_soca`:
> 1317: Test Val : 1.3578991055153833e-01
1317: Ref  Val : 9.2820018030662005e-02
1317: Delta    : 4.2969892520876324e-02
1317: Relative tolerance: 1.1430496429110016e-07
1317: Absolute tolerance: 0.0000000000000000e+00
1317: Test Line: 'RPCGMinimizer: reduction in residual norm = 1.3578991055153833e-01'
1317: Ref Line : 'RPCGMinimizer: reduction in residual norm = 9.2820018030662005e-02'

Seems to be a pretty significant difference, not sure why.

- `test_soca_3dvar_single_ob`:
> 1318: Test Val : 0.0000000000000000e+00
1318: Ref  Val : 2.0050958797774292e-16
1318: Delta    : 2.0050958797774292e-16
1318: Relative tolerance: 1.0025479398887145e-22
1318: Absolute tolerance: 0.0000000000000000e+00
1318: Test Line: 'RPCGMinimizer: reduction in residual norm = 0.0000000000000000e+00'
1318: Ref Line : 'RPCGMinimizer: reduction in residual norm = 2.0050958797774292e-16'

Not sure how significant that is..

- `test_soca_3dvarbump`:
> 1319: Test Val : 1.5167557032312473e-01
1319: Ref  Val : 1.5167539132635138e-01
1319: Delta    : 1.7899677334787256e-07
1319: Relative tolerance: 1.5167548082473804e-07
1319: Absolute tolerance: 0.0000000000000000e+00
1319: Test Line: 'RPCGMinimizer: reduction in residual norm = 1.5167557032312473e-01'
1319: Ref Line : 'RPCGMinimizer: reduction in residual norm = 1.5167539132635138e-01'

Seems like tolerance can be changed.

- `test_soca_3dvar_godas`:
> 1320: Test Val : 1.4657743150452185e-01
1320: Ref  Val : 1.1870594894077013e-01
1320: Delta    : 2.7871482563751715e-02
1320: Relative tolerance: 1.3264169022264599e-07
1320: Absolute tolerance: 0.0000000000000000e+00
1320: Test Line: 'RPCGMinimizer: reduction in residual norm = 1.4657743150452185e-01'
1320: Ref Line : 'RPCGMinimizer: reduction in residual norm = 1.1870594894077013e-01'

Significant difference; not sure why.

- `test_soca_addincrement` and `test_soca_convertincrement`: may depend on the output of the failing tests?

- `test_soca_3dvarlowres_soca`: 
> 1323: Test Val : 1.0675545865616178e-01
1323: Ref  Val : 1.0675524134831732e-01
1323: Delta    : 2.1730784445506135e-07
1323: Relative tolerance: 1.0675535000223953e-07
1323: Absolute tolerance: 0.0000000000000000e+00
1323: Test Line: 'RPCGMinimizer: reduction in residual norm = 1.0675545865616178e-01'
1323: Ref Line : 'RPCGMinimizer: reduction in residual norm = 1.0675524134831732e-01'

Seems like tolerance can be adjusted.

- `test_soca_3dvarfgat`, `test_soca_3dfgat_pseudo`, `test_soca_3dhybfgat` (output not shown for the last two, but similar diffs):
> 1325: Test Val : 4.5208697320818997e-01
1325: Ref  Val : 3.4000878538451623e-01
1325: Delta    : 1.1207818782367374e-01
1325: Relative tolerance: 3.9604787929635310e-07
1325: Absolute tolerance: 0.0000000000000000e+00
1325: Test Line: 'RPCGMinimizer: reduction in residual norm = 4.5208697320818997e-01'
1325: Ref Line : 'RPCGMinimizer: reduction in residual norm = 3.4000878538451623e-01'

Significant difference, not sure why.

- `test_soca_checkpointmodel`:
> 1333: Test Val : 6.0186193018435992e+00
1333: Ref  Val : 6.0197585704204624e+00
1333: Delta    : 1.1392685768631949e-03
1333: Relative tolerance: 6.0191889361320300e-06
1333: Absolute tolerance: 0.0000000000000000e+00
1333: Test Line: '   tocn   min=-1.8883899372702533   max=31.7004645720658580   mean=6.0186193018435992'
1333: Ref Line : '   tocn   min=-1.8883899372702533   max=31.7004645720658580   mean=6.0197585704204624'

Is a difference like this reasonable?

### To Reproduce:
orion; load modules
```
module purge
module use /work/noaa/da/role-da/spack-stack/modulefiles
module load miniconda/3.9.7
module load ecflow/5.8.4
module use /work/noaa/da/role-da/spack-stack/spack-stack-v1/envs/skylab-1.0.0-intel-2022.0.2/install/modulefiles/Core
module load stack-intel/2022.0.2
module load stack-intel-oneapi-mpi/2021.5.1
module load stack-python/3.9.7
module load jedi-ewok-env/1.0.0 jedi-fv3-env/1.0.0 nco/5.0.6
```

### LINKED ISSUE #797: ctest failures on Cheyenne with Intel 19.1.1 for jedi-bundle@1.0.0
## Description
When running the full ctests for jedi-bundle on Cheyenne with Intel-19.1.1 using spack-stack, the following tests fail:
```
	1321 - test_soca_convertstate (Failed)
	1329 - test_soca_create_kmask (Failed)
	1330 - test_soca_enspert (Failed)
	1331 - test_soca_sqrtvertloc (Failed)
	1344 - test_soca_varchange_balance (Failed)
	1345 - test_soca_varchange_balance_TSSSH (Failed)
	1354 - test_soca_ensvariance (Failed)
	1357 - test_soca_ensrecenter (Failed)
	1358 - test_soca_hybridgain (Failed)
	1363 - test_soca_dirac_soca_cov (Failed)
	1364 - test_soca_dirac_socahyb_cov (Failed)
	1371 - test_soca_enshofx (Failed)
	1372 - test_soca_3dvar_soca (Failed)
	1373 - test_soca_3dvar_single_ob (Failed)
	1374 - test_soca_3dvarbump (Failed)
	1375 - test_soca_3dvar_godas (Failed)
	1376 - test_soca_addincrement (Failed)
	1377 - test_soca_convertincrement (Failed)
	1378 - test_soca_3dvarlowres_soca (Failed)
	1380 - test_soca_3dvarfgat (Failed)
	1382 - test_soca_3dhyb (Failed)
	1383 - test_soca_3dhybfgat (Failed)
	1384 - test_soca_letkf (Failed)
	1385 - test_soca_letkf_split_observer (Failed)
	1386 - letkf_split_observer_post (Failed)
	1387 - test_soca_letkf_split_solver (Failed)
	1388 - test_soca_checkpointmodel (Failed)
```
The first test fails due to floating point mismatches, which I think look acceptable - we could change the tolerance to make this test pass. The second test fails due to the usual problem that this Python code needs to be called with `mpiexec -np 1` because of the MPi dependencies in the Python libraries.

I didn't look in detail at the other failures, I assume some are due to the `test_soca_create_kmask` mask test failing.

I am attaching the `log.ctest.rerun` file here.

[log.ctest.rerun.log](https://github.com/JCSDA-internal/soca/files/9269257/log.ctest.rerun.log)

### To Reproduce:
1. Log into Cheyenne
2. Setup the environment
```
module purge
module unuse /glade/u/apps/ch/modulefiles/default/compilers
export MODULEPATH_ROOT=/glade/work/jedipara/cheyenne/spack-stack/modulefiles
module use /glade/work/jedipara/cheyenne/spack-stack/modulefiles/compilers
module use /glade/work/jedipara/cheyenne/spack-stack/modulefiles/misc
module load ecflow/5.8.4
module load miniconda/3.9.12

ulimit -s unlimited
module use /glade/work/jedipara/cheyenne/spack-stack/spack-stack-v1/envs/skylab-1.0.0-intel-19.1.1.217/install/modulefiles/Core
module load stack-intel/19.1.1.217
module load stack-intel-mpi/2019.7.217
module load stack-python/3.9.12

module load jedi-fv3-env/1.0.0
module load jedi-ewok-env/1.0.0
module load nco/5.0.6
```
3. Check out jedi-bundle, build the code, run ctests4. 
```
git clone -b 1.0.0 https://github.com/jcsda-internal/jedi-bundle
mkdir build && cd build
ecbuild ../jedi-bundle 2>&1 | tee log.ecbuild
# Avoid being kicked out of the system ...
nice -n 19 make -j3 VERBOSE=1 2>&1 | tee log.make
ctest -R get_ 2>&1 | tee log.ctest.get
```
4. Hop onto a compute node using `qsub`, then rerun the "Setup Environment steps"
```
qsub -A NMMM0015 -l walltime=08:00:00 -q economy -l select=1:ncpus=36:mpiprocs=36:ompthreads=1
module purge
...
module load nco/5.0.6
```
5. Run `ctest`
```
ctest -E get_ 2>&1 | tee log.ctest.noget
```

## Additional context
This is with `jedi-bundle@1.0.0`. I will try the latest develop version now, just in case some errors have been fixed already.

### PR DESCRIPTION
**Pull request by travissluka**
## Description

Cleanup of the soca ctests to get rid of several problems:
- the test tolerances needed were quite high in some ctests when running across different compilers
- The tests overwrite files create by and/or needed by other tests, so you cannot run the tests out of order or in parallel, and some of the outputs of the tests are lost when the next test is run
- tests were failing on orion and elsewhere

### To address those issues the following were changed:

To reduce the tolerances needed: 
  - The 5 deg 4 member ensemble is pregenerated and added to `test/Data/72x35x25/restart_ens`, a large part of the cross-compiler test differences were due to small differences in the generation of the ensemble by `enspert`, which throws off all the downstream tests.
  - Some of the tests that were only using 1 iteration of var were upped to 5 iterations. It seems the first iteration produces wildly different cost function numbers across the compilers, its not so bad after a couple of iterations.
  - The `3dvarbump` test was wrong, it was using the 5 deg state loaded into the 10 deg geometry... not sure why that didn't throw any runtime warnings... I changed the test to use all low res files.
 - `kmask.nc` is not needed anymore, the tests that generate it have been removed. Instead a pregenerated `Data/72x35x25/dcdt.nc` that contains the ice/ocean jacobian that used to be in kmask has been added
 - I made sure `NOTRAPFPE` was set for the tests that needed them, and only the tests that need them.

To reorganize the tests so that output is better kept
- Each test runs in its own work directory under `test/test_workdir/<test_name>`. We can now properly run tests in parallel without their outputs being overwritten. That working directory gets `data_static`, `data_generated` (see next bullet point)  and the usual `testinput`, `testoutput`, `testref` directories linked in
- the "Data" has been separated into `data_static` (i.e. the data files that are included via git-lfs) and `data_generated/<test_name>` so that within a yaml file we can explicitly see where the data is coming from. E.g. most of the tests have in their geometry section the following, which lets you instantly see that all the data files come from git-lfs, except the gridspec which comes from the `gridgen` ctest:
```yaml
geometry: &geom
    geom_grid_file: data_generated/gridgen/soca_gridspec.72x35x25.nc
    mom6_input_nml: data_static/72x35x25/input.nml
    fields metadata: data_static/fields_metadata.yml
```
-  All tests have their generated output placed in `data_generated/<test_name>` so we can keep track of each tests' output better. Before, for example, multiple tests were producing `sst.out.nc` files, which got overwritten. Now all tests have output in unique directories.

In order to be able to run each test in its own directory without having to symlink in a whole bunch of `data_static/` files, 
- the paths to certain data files (`rossrad.dat`,`godas_sst_bgerr.nc`) can be specified in the yaml. If no yaml parameter is given, the default is set to use the currently used hardcoded path.

Other small things: 
- get rid of the annoying `WARNING from PE    0: Unused line in MOM_input` that were cluttering up the output log
- made sure all the test dependencies in `CMakeLists.txt` were up to date (some were wrong and/or missing)
- the single ob 3dvar test had to be disabled because it was failing on the latest intel compiler.... need to take a closer look at some point to figure out why.

### Issue(s) addressed
- closes  #797 
- closes #773 
- partially addresses #784 , still need to remove the code that used it
- 
## Testing

Tested on the following, and the tolerances have been set so that they pass (highest tolerance is ~ 1e-5, so much better than it used to be!)

- personal computer: gnu 9.x
- orion with gcc 10.2
- orion with older intel (intel 2020)
- orion with intel spack stack (intel with 2022)

The test references were all regenerated using GCC 10.2 on orion.

### CODE REVIEW

**Comment by travissluka** [comment]
@climbfuji I'll try to find time tomorrow to remember how to log onto Cheyenne and see if the ctests pass now. But if you happen to beat me to it, save your logs and i'll see if I need to up any more tolerances

File: test/Data/72x35x25/MOM_input [hunk 1]
```diff
@@ -588,3 +588,5 @@ FLUXCONST = 0.5                 !   [m day-1]
 ! === module MOM_restart ===
 
 ! === module MOM_file_parser ===
+REPORT_UNUSED_PARAMS = False     !   [Boolean] default = True
```
Comment (guillaumevernieres) [inline test/Data/72x35x25/MOM_input:591]: :tada:

**Comment by climbfuji** [comment]
> @climbfuji I'll try to find time tomorrow to remember how to log onto Cheyenne and see if the ctests pass now. But if you happen to beat me to it, save your logs and i'll see if I need to up any more tolerances

I can run the tests in my clang container as well as on Cheyenne.

**Review by guillaumevernieres — APPROVED** [review]
Nice cleanup. I browsed through the changes, I didn't see anything wrong but ... 160 files!
I like the `REPORT_UNUSED_PARAMS` option!
Good catch on the 3dvarbump, no clue why it worked.

**Review by kbhargava — APPROVED** [review]
Yayy!!! FInally all soca ctests pass on orion! Though I had 1 ctest `test_soca_enshofx `fail the first time I ran `ctest`. I tried again using` ctest -v -R test_soca_enshofx `and it passed.

**Comment by liuxiao37k** [comment]
@travissluka @guillaumevernieres I did notice the inconsistency between the state and geometry dimensions in `3dvarbump` last time I did a PR, and actually played with it a little bit. Here are my observations:
1) Yes, you can indeed load a higher res state into a lower res geometry without triggering an error, but I THINK the coordinates would be wrong. Evidence:
when there was such an inconsistency, a large majority of the obs were removed by first guess (background check); after loading the state of correct size, much less got removed by first guess. All other obs-based filters were not affected.
2) I THINK, the reason for 1) is that a corner of the state [0 NIGLOBAL], [0 NJGLOBAL] is loaded in when the dimension of state is larger than geometry. As long as this can be executed, no error is reported. Evidence:
If I tried to load a lower res state into a higher res geometry, I would get the following error:
```
FATAL from PE     0: NETCDF ERROR: NetCDF: Start+count exceeds dimension bound File=./INPUT_small/MOM.res.nc
```
`Start+count` is probably just checking [0 NIGLOBAL], [0 NJGLOBAL].

**Review by liuxiao37k — COMMENTED** [review]
Tested several times on Orion and observed some behaviors:
1) `test_soca_enshofx` had a tendency to fail for the 1st time, and always succeeded after the 1st time. Consistent with what @kbhargava said above. Error was not related to tolerance.
2) Unless `test_soca_gridgen` was run first, other tests often would fail even though `TEST_DEPENDS test_soca_gridgen` is set.
I believe both 1) and 2) had something to do with tonight's (just tonight?) filesystem on Orion. It seems the required file (gridspec for 1 and ocean.stats for 2) was not fully ready yet before one of the processors tried to access/write into it. That probably explains why running the test after the 1st time was always successful (file already existed).
3) `test_soca_convertate` failed a couple times due to tolerance.
Overall nice cleanup!!!

**Comment by frolovsa** [comment]
Impressive clean up @travissluka

**Comment by climbfuji** [comment]
@travissluka
1. On Cheyenne with GNU, all SOCA tests passed
2. On Cheyenne with Intel, the following SOCA test failed first time, all others passed. I reran it and it passed in 5.5s
```
The following tests FAILED:
	1369 - test_soca_enshofx (Timeout)
```
3. In my clang container on aarch64 Ubuntu 20.04, the following SOCA tests failed:
```
	1272 - test_soca_ensvariance (Failed)
	1279 - test_soca_dirac_soca_cov (Failed)
	1280 - test_soca_dirac_socahyb_cov (Failed)
	1290 - test_soca_3dvarbump (Failed)
	1296 - test_soca_3dhybfgat (Failed)
	1301 - test_soca_convertincrement (Failed)
```
Example for 1272:
```
OOPS_STATS Run end                                  - Runtime:      0.26 sec,  Memory: total:   236.03 Mb, per task: min =   117.75 Mb, max =   118.28 Mb
Run: Finishing oops::EnsVariance<SOCA> with status = 0
terminate called after throwing an instance of 'oops::TestReferenceFloatMismatchError'
  what():  Test reference Float mismatch @ Line:10
Test Val : 1.4722851834880000e-03
Ref  Val : 1.4722763694402000e-03
Delta    : 8.8140477999976846e-09
Relative tolerance: 1.4722807764640998e-09
Absolute tolerance: 0.0000000000000000e+00
Test Line: '    ssh   min=0.0000000000000000   max=0.0014722851834880   mean=0.0000180190538505'
Ref Line : '    ssh   min=0.0000000000000000   max=0.0014722763694402   mean=0.0000180187484047'
```

**Comment by travissluka** [comment]
@climbfuji I've blindly increased tolerance for the tests you mentioned in the hope that it fixes your fails... since I don't have access to an aarch64 machine. I'm going to go ahead and merge this PR today once I have a companion PR to soca-science. We can deal with any remaining failures you see in a followup PR

**Review by climbfuji — APPROVED** [review]
Thanks for adjusting the tolerances. We'll retest with the new containers once this PR is merged.

**Pull request merged**